lib/elixir/lib/calendar/date/range.ex

```diff
@@ -0,0 +1,90 @@
+defmodule Date.Range do
+  @moduledoc """
+  Defines a range of dates.
+
+  A range of dates represents a discrete number of dates where
+  the first and last values are dates with matching calendars.
+
+  Ranges of dates can be either increasing (`first <= last`) or
+  decreasing (`first > last`). They are also always inclusive.
+
+  A range of dates implements the `Enumerable` protocol, which means
+  functions in the `Enum` module can be used to work with
+  ranges:
+
+      iex> range = Date.range(~D[2001-01-01], ~D[2002-01-01])
+      iex> Enum.count(range)
+      366
+      iex> Enum.member?(range, ~D[2001-02-01])
+      true
+      iex> Enum.reduce(range, 0, fn(_date, acc) -> acc - 1 end)
+      -366
+
+  """
+
+  defstruct [:first, :last, :kind, :first_rata_die, :last_rata_die]
+
+  def new(%{calendar: fc} = first, %{calendar: lc} = last) when fc == lc do
+    %Date.Range{
+      first: first,
+      last: last,
+      first_rata_die: Date.to_rata_die(first),
+      last_rata_die: Date.to_rata_die(last),
+      kind: get_range_kind(first, last),
+    }
+  end
+  def new(_, _) do
+    raise ArgumentError,
+      "Both dates must have matching calendars"
+  end
+
+  defp get_range_kind(first, last) when first > last, do: :desc
+  defp get_range_kind(_first, _last), do: :asc
+
+  defimpl Enumerable, for: Date.Range do
+    def member?(range, date) do
```

Comment:
For reference, here is how I would implement `member?`:

```elixir
def member?(%{first: %{calendar: calendar, year: year1, month: month1, day: day1},
              last: %{calendar: calendar, year: year2, month: month2, day: day2},
              first_rata_die: first_rata_die, last_rata_die: last_rata_die},
            %Date{calendar: calendar, year: year, month: month, day: day}
  date1 = {year1, month1, day1}
  date2 = {year2, month2, day2}
  date = {year, month, day}

  if first_rata_die <= last_rata_die do
    {:ok, date >= date1 and date <= date2}
  else
    {:ok, date >= date2 and date <= date1}
  end
end
def member?(_, _) do
  {:ok, false}
end
```

Since range assumes same calendar, we don't need the rata die conversion.